GitUI/CommandsDialogs/FormCommit.cs

```diff
@@ -155,10 +155,13 @@ public sealed partial class FormCommit : GitModuleForm
         private readonly AsyncLoader _unstagedLoader = new AsyncLoader();
         private readonly bool _useFormCommitMessage = AppSettings.UseFormCommitMessage;
         private readonly CancellationTokenSequence _interactiveAddSequence = new CancellationTokenSequence();
+        private readonly CancellationTokenSequence _cancellationTokenSequence = new CancellationTokenSequence();
         private readonly SplitterManager _splitterManager = new SplitterManager(new AppSettingsPath("CommitDialog"));
         private readonly Subject<string> _selectionFilterSubject = new Subject<string>();
         private readonly IFullPathResolver _fullPathResolver;
         private readonly List<string> _formattedLines = new List<string>();
+        private readonly JoinableTaskCollection _joinableTasks;
+        private readonly JoinableTaskFactory _joinableTaskFactory;
```

Comment:
❗️ We need the tasks to be in the main task collection, and there's no way to create a factory that is backed by two collections. This means we need to use the original factory and manually add the resulting tasks to `_joinableTasks`.

Reply:
Can you file this use case on vs-threading? We recently picked up the ability to nest collections in JTF, but have not exposed this. It sounds like it would be useful to you.

Reply:
Do I need to change the code in mean time, or what I have now is passable?

Reply:
We need to change the code. Our test cleanup sequence relies on asynchronous operations getting tracked in the main collection. Failing to add operations there could lead to cross-test interference.

https://github.com/gitextensions/gitextensions/blob/96090c8d0114397fb0bcfbbb342261d1da45e904/UnitTests/CommonTestUtils/ConfigureJoinableTaskFactoryAttribute.cs#L53-L57

Reply:
I think I'm going mental... Clearly I'm doing something wrong.

I've tried changing the main code back to use `ThreadHelper.JoinableTaskFactory.RunAsync().FileAndForget()` and then changed the test code to use `ThreadHelper.JoinableTaskFactory.Run(() => ThreadHelper.JoinPendingOperationsAsync())`. 
```diff
diff --git a/GitUI/CommandsDialogs/FormCommit.cs b/GitUI/CommandsDialogs/FormCommit.cs
index ac0526643..3ff1d4162 100644
--- a/GitUI/CommandsDialogs/FormCommit.cs
+++ b/GitUI/CommandsDialogs/FormCommit.cs
@@ -427,5 +427,5 @@ protected override void OnShown(EventArgs e)
 
             var token = _updateBranchNameDisplaySequence.Next();
-            _joinableTaskFactory.RunAsync(() => UpdateBranchNameDisplayAsync(token)).FileAndForget();
+            ThreadHelper.JoinableTaskFactory.RunAsync(() => UpdateBranchNameDisplayAsync(token)).FileAndForget();
             UpdateAuthorInfo();
 
@@ -1195,5 +1195,5 @@ private void ShowChanges(GitItemStatus item, bool staged)
             }
 
-            _joinableTaskFactory.RunAsync(async () =>
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
             {
                 await SetSelectedDiffAsync(item, staged);
@@ -3229,5 +3229,5 @@ private void createBranchToolStripButton_Click(object sender, EventArgs e)
 
             var token = _updateBranchNameDisplaySequence.Next();
-            _joinableTaskFactory.RunAsync(() => UpdateBranchNameDisplayAsync(token)).FileAndForget();
+            ThreadHelper.JoinableTaskFactory.RunAsync(() => UpdateBranchNameDisplayAsync(token)).FileAndForget();
         }
``` 

This way the background tasks should be added to the main collection and we could await all of them. Or so I thought.

However I get a deadlock in `UpdateBranchNameDisplayAsync_should_display_branch_and_no_remote_info_in_statusbar` test.
```diff
diff --git a/UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs b/UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs
index 121c04a74..a8b961c9f 100644
--- a/UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs
+++ b/UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs
@@ -96,5 +96,5 @@ public void UpdateBranchNameDisplayAsync_should_display_branch_and_no_remote_inf
             RunFormTest(form =>
             {
-                ThreadHelper.JoinableTaskFactory.Run(() => form.GetTestAccessor().RunningTasks.JoinTillEmptyAsync());
+                ThreadHelper.JoinableTaskFactory.Run(() => ThreadHelper.JoinPendingOperationsAsync());
 
                 var currentBranchNameLabelStatus = form.GetTestAccessor().CurrentBranchNameLabelStatus;
```
And I can't seem to figure out what causes it...

@AArnott is there a way to see what tasks are in a collection? I've tried going through a debug window but I couldn't see anything meaningful.

Reply:
`ThreadHelper.JoinableTaskFactory.Run(() => ThreadHelper.JoinPendingOperationsAsync());` will probably deadlock all on its own if you try it. Assuming `ThreadHelper.JoinPendingOperationsAsync()` calls `JoinTillEmptyAsync()` on its associated collection, you're waiting on yourself to finish, which will never happen. You need to be careful to _not_ use a JTF instance to wait for completion of its own collection because the waiting is itself in that same collection. Try this instead:

`ThreadHelper.JoinableTaskContext.Factory.Run(() => ThreadHelper.JoinPendingOperationsAsync());`